In [69]:
import pandas as pd
df = pd.read_csv('clustered_gd_noLSH.csv',header=0,index_col=False) 
df = df.set_index('user_id')
df = df.drop('Clusters5', 1)
df.head()
#df = pd.DataFrame(X, columns=['f{}'.format(i) for i in range(8)])

C:\Users\bkuro\AppData\Local\Temp\ipykernel_21420\3934716007.py:4: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  df = df.drop('Clusters5', 1)


churches  resorts  beaches  parks  theatres  museums  malls   zoo  \
user_id                                                                      
User 1        0.0      0.0     3.63   3.65       5.0     2.92    5.0  2.35   
User 2        0.0      0.0     3.63   3.65       5.0     2.92    5.0  2.64   
User 3        0.0      0.0     3.63   3.63       5.0     2.92    5.0  2.64   
User 4        0.0      0.5     3.63   3.63       5.0     2.92    5.0  2.35   
User 5        0.0      0.0     3.63   3.63       5.0     2.92    5.0  2.64   

         restaurants  pubs-bars  ...  dance-clubs  swimming-pools  gyms  \
user_id                          ...                                      
User 1          2.33       2.64  ...         0.59             0.5   0.0   
User 2          2.33       2.65  ...         0.59             0.5   0.0   
User 3          2.33       2.64  ...         0.59             0.5   0.0   
User 4          2.33       2.64  ...         0.59             0.5   0.0   
User 5          2.33       2.64  ...         0.59             0.5   0.0   

         bakeries  beauty-spas  cafes  view-points  monuments  gardens  \
user_id                                                                  
User 1        0.5          0.0    0.0          0.0        0.0      0.0   
User 2        0.5          0.0    0.0          0.0        0.0      0.0   
User 3        0.5          0.0    0.0          0.0        0.0      0.0   
User 4        0.5          0.0    0.0          0.0        0.0      0.0   
User 5        0.5          0.0    0.0          0.0        0.0      0.0   

         Clusters4  
user_id             
User 1           1  
User 2           1  
User 3           1  
User 4           1  
User 5           1  

[5 rows x 23 columns]

In [70]:
#df = df.drop(columns=["Clusters4"])

In [71]:
y = df["Clusters4"]

In [72]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(df, y, test_size=0.25, random_state=90210)

In [73]:
# to divide the df by cluster, we need to ensure we use the correct class labels, we'll use pandas to do that
train_clusters = X_train.copy()
test_clusters = X_test.copy()
train_clusters['y'] = y_train
test_clusters['y'] = y_test


# locate the "0" cluster
train_0 = train_clusters.loc[train_clusters.Clusters4 == 0] 
test_0 = test_clusters.loc[test_clusters.Clusters4 == 0]
y_train_0 = train_0.y.values
y_test_0 = test_0.y.values
# locate the "1" cluster
train_1 = train_clusters.loc[train_clusters.Clusters4 == 1] 
test_1 = test_clusters.loc[test_clusters.Clusters4 == 1]
y_train_1 = train_1.y.values
y_test_1 = test_1.y.values
# locate the "2" cluster
train_2 = train_clusters.loc[train_clusters.Clusters4 == 2] 
test_2 = test_clusters.loc[test_clusters.Clusters4 == 2]
y_train_2 = train_2.y.values
y_test_2 = test_2.y.values
# locate the "3" cluster
train_3 = train_clusters.loc[train_clusters.Clusters4 == 3] 
test_3 = test_clusters.loc[test_clusters.Clusters4 == 3]
y_train_3 = train_3.y.values
y_test_3 = test_3.y.values


In [74]:
test_3

churches  resorts  beaches  parks  theatres  museums  malls   zoo  \
user_id                                                                        
User 3468      0.00     0.00     1.48   1.41      1.40     1.41   1.42  1.45   
User 2480      0.79     1.42     1.40   1.39      1.41     1.61   1.45  2.64   
User 4388      1.47     1.65     2.46   1.89      1.88     1.88   1.91  1.92   
User 1659      1.50     1.63     1.99   1.95      1.94     2.05   5.00  2.12   
User 4290      0.00     1.46     1.44   1.43      1.44     1.49   1.49  2.70   
...             ...      ...      ...    ...       ...      ...    ...   ...   
User 2527      0.76     0.79     1.57   1.55      1.54     1.55   5.00  1.59   
User 1931      0.72     5.00     1.37   1.38      1.44     1.45   5.00  4.13   
User 679       0.00     0.00     1.48   1.43      1.42     1.44   1.48  1.48   
User 4911      1.14     1.32     2.64   2.64      2.75     2.79   5.00  3.12   
User 142       0.78     0.72     1.27   1.29      1.36     1.38   5.00  1.41   

           restaurants  pubs-bars  ...  swimming-pools  gyms  bakeries  \
user_id                            ...                                   
User 3468         2.66       3.28  ...            0.84  0.86      1.04   
User 2480         3.30       3.27  ...            0.86  1.04      1.06   
User 4388         2.49       2.28  ...            5.00  0.00      0.00   
User 1659         2.62       2.23  ...            5.00  0.00      0.00   
User 4290         3.33       3.32  ...            1.01  1.00      1.02   
...                ...        ...  ...             ...   ...       ...   
User 2527         2.44       3.03  ...            0.70  0.71      0.73   
User 1931         5.00       5.00  ...            0.66  0.67      0.70   
User 679          2.71       2.73  ...            0.78  0.80      0.92   
User 4911         3.11       2.57  ...            0.67  0.67      0.69   
User 142          5.00       5.00  ...            0.75  0.75      0.85   

           beauty-spas  cafes  view-points  monuments  gardens  Clusters4  y  
user_id                                                                       
User 3468         1.11   0.00         0.00       0.00     0.00          3  3  
User 2480         0.54   0.00         0.00       0.00     0.00          3  3  
User 4388         0.61   1.48         1.46       1.45     1.45          3  3  
User 1659         0.58   1.50         1.49       1.47     1.49          3  3  
User 4290         0.00   0.00         0.00       0.00     1.59          3  3  
...                ...    ...          ...        ...      ...        ... ..  
User 2527         0.75   0.76         0.75       0.74     0.74          3  3  
User 1931         0.72   0.73         0.72       0.70     0.74          3  3  
User 679          0.00   0.00         0.00       0.00     0.00          3  3  
User 4911         0.00   0.00         1.03       1.04     1.06          3  3  
User 142          0.80   0.68         0.66       0.65     0.66          3  3  

[333 rows x 24 columns]

In [76]:
# the base dataset has no "clusters" feature
X_train_base = X_train.drop(columns=['Clusters4'])
X_test_base = X_test.drop(columns=['Clusters4'])
# drop the targets from the training set
X_train_0 = train_0.drop(columns=['y'])
X_test_0 = test_0.drop(columns=['y'])
X_train_1 = train_1.drop(columns=['y'])
X_test_1 = test_1.drop(columns=['y'])
X_train_2 = train_2.drop(columns=['y'])
X_test_2 = test_2.drop(columns=['y'])
X_train_3 = train_3.drop(columns=['y'])
X_test_3 = test_3.drop(columns=['y'])
datasets = {
    'base': (X_train_base, y_train, X_test_base, y_test), }
#'cluster-feature': (X_train, y_train, X_test, y_test), 
#'cluster-0': (X_train_0, y_train_0, X_test_0, y_test_0),
#'cluster-1': (X_train_1, y_train_1, X_test_1, y_test_1),
#'cluster-2': (X_train_2, y_train_2, X_test_2, y_test_2),
#'cluster-3': (X_train_3, y_train_3, X_test_3, y_test_3),


In [77]:
y_test.head()

user_id
User 3468    3
User 2480    3
User 4388    3
User 4196    1
User 2224    1
Name: Clusters4, dtype: int64

In [ ]:
from numpy import mean
from numpy import std
from sklearn.datasets import make_classification
from sklearn.model_selection import RepeatedKFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn import model_selection
from sklearn.metrics import classification_report

In [78]:
classifier = LogisticRegression(solver='lbfgs',
                                max_iter=800,
                                random_state=1)
classifier.fit(X_train_base, y_train)

LogisticRegression(max_iter=800, random_state=1)

In [79]:
y_pred = classifier.predict(X_test_base)
results = pd.DataFrame({"Prediction": y_pred, "Actual": y_test}).reset_index(drop=True)
results.head(10)

Prediction  Actual
0           3       3
1           3       3
2           3       3
3           1       1
4           1       1
5           3       3
6           0       0
7           0       0
8           0       0
9           0       0

In [85]:
# evaluate a logistic regression model using repeated k-fold cross-validation

# prepare the cross-validation procedure
rcv = RepeatedKFold(n_splits=10, n_repeats=5, random_state=1)
scoring = ['accuracy','precision_weighted','recall_weighted','f1_weighted']
# create model
model = LogisticRegression() #classifier
# evaluate model
scores = cross_val_score(model, X_train_base, y_train, scoring='accuracy', cv=rcv, n_jobs=-1)

# report performance
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))
# print(condition)

0.9875366568914956
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       309
           1       0.99      0.99      0.99       467
           2       0.98      0.99      0.98       255
           3       0.99      0.98      0.99       333

    accuracy                           0.99      1364
   macro avg       0.99      0.99      0.99      1364
weighted avg       0.99      0.99      0.99      1364






def run_exps(datasets: dict) -> pd.DataFrame:
    '''
    runs experiments on a dict of datasets
    '''
    # initialize a logistic regression classifier
    model = LogisticRegression(class_weight='balanced', solver='lbfgs', random_state=997, max_iter=800)
    
    dfs = []
    results = []
    conditions = []
    scoring = ['accuracy','precision_weighted','recall_weighted','f1_weighted']

    for condition, splits in datasets.items():
            X_train = splits[0]
            y_train = splits[1]
            X_test = splits[2]
            y_test = splits[3]

            kfold = model_selection.KFold(n_splits=5, shuffle=True, random_state=90210)
            cv_results = model_selection.cross_validate(model, X_train, y_train, cv=kfold, scoring=scoring)
            clf = model.fit(X_train, y_train)
            y_pred = clf.predict(X_test)
            print(condition)
            print(classification_report(y_test, y_pred))

    results.append(cv_results)
    conditions.append(condition)

    this_df = pd.DataFrame(cv_results)
    this_df['condition'] = condition
    dfs.append(this_df)

    final = pd.concat(dfs, ignore_index=True)
    
    # We have wide format data, lets use pd.melt to fix this
    results_long = pd.melt(final,id_vars=['condition'],var_name='metrics', value_name='values')
    
    # fit time metrics, we don't need these
    time_metrics = ['fit_time','score_time'] 
    results = results_long[~results_long['metrics'].isin(time_metrics)] # get df without fit data
    results = results.sort_values(by='values')
    
    return results

df = run_exps(datasets)

import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
plt.figure(figsize=(20, 12))
sns.set(font_scale=2.5)
g = sns.boxplot(x="condition", y="values", hue="metrics", data=df, palette="Set3")
plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.)
plt.title('Comparison of Dataset by Classification Metric')

pd.pivot_table(df, index='condition',columns=['metrics'],values=['values'], aggfunc='mean')